msvc/scripts/genconsts.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using System.IO;
+using System.Reflection;
+using System.Diagnostics;
+using System.Text;
+using System.Text.RegularExpressions;
+
+public static class Program {
+    public static int Main (string[] args) {
+        var myAssembly = Assembly.GetExecutingAssembly ();
+        var codeBase = new Uri (myAssembly.CodeBase);
+        var executablePath = Path.GetFullPath (codeBase.LocalPath);
+        var executableDirectory = Path.GetDirectoryName (executablePath);
+
+        var winsetupDirectory = Path.Combine (executableDirectory, "..");
+        var winsetupPath = Path.Combine (winsetupDirectory, "winsetup.bat");
+
+        var psi = new ProcessStartInfo (winsetupPath) {
+            WorkingDirectory = winsetupDirectory,
+            UseShellExecute = false,
+            ErrorDialog = false,
+            RedirectStandardOutput = true
+        };
+
+        string monoVersion, monoCorlibVersion;
+
+        Process winsetupProcess;
+
+        try {
+            winsetupProcess = Process.Start (psi);
+        } catch (Exception exc) {
+            Console.Error.WriteLine ("Failed starting winsetup.bat");
+            Console.Error.WriteLine (exc);
+            return 1;
+        }
+
+        using (winsetupProcess) {
+            var outputBuffer = new StringBuilder ();
+
+            winsetupProcess.OutputDataReceived += (s, e) => {
+                outputBuffer.AppendLine (e.Data);
+            };
+            winsetupProcess.BeginOutputReadLine ();
```

Comment:
You should also be aware, at least for future, that when you run stuff,  and don't read both stderr and stdout, that you risk deadlock. I've seen it many times. In this case you are probably ok, probably no writes to stderr.

In native code it is quite messy. C#/BCL might be easier.
In  native code you have two choices for deadlock-free reading of a process's output.

1. You need to read stdout and stderr on separate threads or 
2. You need to set stdout==stderr, i.e. redirect them to the same handle.

Again, I've seen this deadlock countless times, but it tended to be in C++ and script.exe/jscript, perhaps not C#.

Reply:
It should be OK because I didn't redirect stderr, I think.